lib/clib/Corres_UL_C.thy

```diff
@@ -288,6 +288,43 @@ lemma ccorres_from_vcg_nofail:
   apply (rule hoare_complete, simp add: HoarePartialDef.valid_def)
   done
 
+lemma ccorres_to_vcg_with_prop:
+  "\<lbrakk>ccorres_underlying srel \<Gamma> rrel xf arrel axf P P' [] a c; no_fail Q a; \<lbrace>P\<rbrace> a \<lbrace>R\<rbrace>\<rbrakk>
+   \<Longrightarrow> \<Gamma> \<turnstile> {s'. P s \<and> Q s \<and> s' \<in> P' \<and> (s, s') \<in> srel}
+           c {t'. \<exists>(rv, t) \<in> fst (a s). (t, t') \<in> srel \<and> rrel rv (xf t') \<and> R rv t}"
+  apply (clarsimp simp: no_fail_def)
+  apply (drule_tac x=s in spec)
+  apply (rule hoare_complete)
+  apply (simp add: HoarePartialDef.valid_def cvalid_def)
+  apply (intro impI allI)
+  apply clarsimp
+  apply (frule (5) ccorres_empty_handler_abrupt)
+  apply (erule (4) ccorresE)
+   apply (erule (1) EHOther)
+  apply (rule image_eqI)
+   apply simp
+  apply (fastforce simp: unif_rrel_simps valid_def)
+  done
```

Comment:
This proof appears nearly identical to `ccorres_to_vcg`. Does it mean that it becomes an instance of this rule? No sense repeating the proof in that case, suggest changing proof of `ccorres_to_vcg` to be an instantiation.
Also, this proof looks a bit ancient (have a feeling that fastforce/force/auto can take it from the `image_eqI` line for example), if you have time maybe try shorten it a bit?

Reply:
Also same comment as Gerwin on the while rule, `clarsimp` includes doing `allI` and `impI` already, so that `simp, intro, clarsimp` might be compressible.

Reply:
Sorry, I meant to make a comment about these rules. There are some variations on this rule: `ccorres_to_vcg'`, which appears in CLib and is pretty much `ccorres_to_vcg` in clarsimp normal form, and then there's also `ccorres_to_vcg_nf` in CRefine, which uses `no_fail` rather than `snd _`. Should I consolidate these? I could prove the `with_prop` version, and get any other variant that's required from it

Reply:
I honestly don't know which of these is the better form. Maybe @lsf37 does. Thumbs up for consolidation if you have time for it.

Reply:
I'd like to hear Gerwin's opinion, but I'm pretty sure that having `ccorres_to_vcg_with_prop` as in this PR, and then replacing the current `ccorres_to_vcg` with

```
lemma ccorres_to_vcg:
  "⟦ccorres_underlying srel Γ rrel xf arrel axf P P' [] a c; no_fail Q a⟧
   ⟹ Γ ⊢ {s'. P s ∧ Q s ∧ s' ∈ P' ∧ (s, s') ∈ srel}
           c {t'. ∃(rv, t) ∈ fst (a s). (t, t') ∈ srel ∧ rrel rv (xf t')}"
  apply (frule (1) ccorres_to_vcg_with_prop[where R="⊤⊤"])
   apply wpsimp
  apply fastforce
  done
```

would be the best way to go.

The original `ccorres_to_vcg` isn't in clarsimp normal form. Then `ccorres_to_vcg'` requires `no_fail` for the abstract function with the precondition `\top`, which is stronger than what I require with this new formulation. And then `ccorres_to_vcg_nf` requires the precondition for `no_fail` to be no stronger than the abstract guard of the `ccorres` rule, which is not necessary.

Reply:
That's not quite right. I suppose it's that `ccorres_to_vcg'` requires the abstract function to not fail on the given abstract state. But non-failure is probably best dealt with via an actual `no_fail` assumption, which this rule doesn't do.

Reply:
It seems like the variants of `ccorres_to_vcg` were actually unused in CRefine. Updates in the last commit

Reply:
Cool. I'm happy with the version in the last commit.